File: prep/implementation/iterations/iteration_cpp/counting_sort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void display(const vector<int> &a) {
  for (int e : a) {
    cout << e << " ";
  }
  cout << "\n";
}

void sort(vector<int> &a, int n) {
  int minval = a[0], maxval = a[0];
  for (int i = 1; i < n; i++) {
    minval = min(minval, a[i]);
    maxval = max(maxval, a[i]);
  }

  int k = abs(maxval - minval) + 1;
  vector<int> b(k);

  for (int e : a) {
    b[e - minval]++;
  }

  int idx = 0;
  for (int i = 0; i < k; i++) {
    int f = b[i];
    while (f--) {
      a[idx++] = i + minval;
    }
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n;
  cin >> n;

  vector<int> a(n);

  for (int i = 0; i < n; i++) {
    cin >> a[i];
  }

  display(a);
  sort(a, n);
  display(a);

  return 0;
}

File: prep/implementation/practice_iterations/main_1/CountingSort.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class CountingSort {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();

      int[] a = new int[n];
      for (int i = 0; i < n; i++) {
        a[i] = in.nextInt();
      }

      display(a, n);
      sort(a, n);
      display(a, n);
    }
  }

  // O(n + k) / O(n + k)
  private static void sort(int[] a, int n) {
    int minVal = a[0];
    int maxVal = a[0];

    for (int i = 1; i < n; i++) {
      minVal = Math.min(minVal, a[i]);
      maxVal = Math.max(maxVal, a[i]);
    }

    int k = maxVal - minVal + 1;
    int[] b = new int[k + 1];
    for (int e : a) {
      b[e - minVal]++;
    }

    int idx = 0;
    for (int i = 0; i < k; i++) {
      int f = b[i];
      while (f-- > 0) {
        a[idx++] = i + minVal;
      }
    }
  }

  private static void swap(int[] a, int x, int y) {
    int t = a[x];
    a[x] = a[y];
    a[y] = t;
  }

  private static void display(int[] a, int n) {
    for (int i = 0; i < n; i++) {
      System.out.printf("%d ", a[i]);
    }
    System.out.println();
  }
}